### PR TITLE
Frida mode for AFL/AFLFast

### PR DESCRIPTION
**Pull request by mkYYY**
## Type of PR

- [x] Changes related to the roadmap (e.g., TODO.md) (type: A) -> Create an issue corresponding to the PR in advance, and refer to this PR in the issue.
- Changes that are not related to the roadmap
    - [ ] Change with multiple possible solutions to the issue (type: B-1) -> Create an issue corresponding to the PR in advance and refer to this PR in the issue.
    - [ ] Change with a single solution (type: B-2) -> There is no need to create an issue corresponding to the PR in advance. Please discuss it in this PR.

## Importance of PR

- Importance of the issue
    - [ ] Large (based on several days to weeks of discussion and verification, e.g., this issue is a blocking issue for other issues on the roadmap, etc.)
    - [x] Medium (based on a few hours to a day of discussion and verification, e.g., this issue is a blocking issue for another minor issue)
    - [ ] Small (apparent changes such as build error)
- Complexity of the solution (code, tests, etc.)
    - [ ] Large (requires several days to several weeks of review)
    - [x] Medium (requires several hours to a day of review)
    - [ ] Small (trivial changes, such as build error)

## PR Overview

This PR adds frida mode for fuzzuf's AFL and AFLFast (possibly AFL-based fuzzers implemented in the future too). 
Users first clone and build `afl-frida-trace.so` in AFLpluspus submodule by `cmake` command with `-DENABLE_FRIDA_TRACE=1`. 
They then specify `--frida=1` on fuzzuf command execution. 

## Concerns (Optional)

- [ ] Performance
- [x] Source Code Quality

- The frida mode functionality introduced by this PR is limited compared to the original AFLplusplus's.
- As @K-atc spotted, frida mode is currently only available through fuzzuf's CLI (i.e. not available through `libfuzzuf.so`).
- Is the document explanatory enough?

---

> The PR author should fill in the following checklist when submitting this PR.

#### Optional Entries
- [ ] If this PR is a PR type A/B-1, there is a cross-link between this PR and the related issue.

#### Mandatory Entries
- [x] The PR title is a summary of the changes.
- [x] Completed each required field of the PR.

---
> The PR author should fill out the following checklist in the comments to confirm that this PR is ready to be merged

- [ ] CI is green or confirmed test run results.
- [ ] All change suggestions from reviewers have been resolved (fixed or foregone).

---
> The maintainer of this repository will set up a reviewer for each PR.
> PR reviewers should review this PR in terms of the checklist below before moving on to a detailed code review. Please comment on their initial response by filling in the checklist below.

#### Optional Entries
- [ ] The reviewer assigned more reviewers if needed.
- [ ] The reviewer noted that it is necessary to break out some of the changes in this PR into other PRs if needed.
- [ ] The reviewer noted that the initial response is insufficient if needed.

#### Mandatory Entries
- [ ] The title of this PR summarizes the changes made by this PR properly.
- [ ] The target branch of this PR is as intended.
- [ ] The reviewer understands the issues in this PR.
- [ ] The reviewer plans to review with an appropriate workload based on the importance of this PR.

---
> When the PR reviewer concludes that this PR is ready to be merged, please fill in the checklist below by posting it in the comment. If there is more than one reviewer, please do this on your own.

#### Optional Entries
- [ ] The reviewer noted that if you believe that new tests are needed to evaluate this PR, they have been noted.
- [ ] If minor refactorings are not mentioned in the PR, I understand the intent.
- [ ] If this PR is a PR type A/B-1, we have agreed on this PR's design, direction, and granularity in the related issue.

#### Mandatory Entries
- [ ] The reviewer understands how this PR addresses the issue and the specific changes.
- [ ] This PR uses the best possible issue resolution that the reviewer can think of.
- [ ] This PR is ready to be merged.

### CODE REVIEW

File: include/fuzzuf/cli/fuzzer/afl/build_afl_fuzzer_from_args.hpp [hunk 1]
```diff
@@ -92,6 +97,18 @@ std::unique_ptr<TFuzzer> BuildAFLFuzzerFromArgs(
...
+    if (afl_options.frida_mode) {
+        setenv("__AFL_DEFER_FORKSRV", "1", 1);
+        fs::path fuzzuf_bin(fuzzer_args.argv[0]);
+        fs::path frida_bin = fuzzuf_bin.parent_path() / "afl-frida-trace.so";
+        setenv("LD_PRELOAD", frida_bin.c_str(), 1);
+
+        struct stat statbuf;
+        stat(frida_bin.c_str(), &statbuf);
+        extra_mem += statbuf.st_size;
+    }
```
Comment (K-atc) [inline include/fuzzuf/cli/fuzzer/afl/build_afl_fuzzer_from_args.hpp]: @mkYYY 
〔本PRのスコープ〕この実装だと **CLIでしか** Frida modeが使えないということになります。（i.e. たとえばfuzzufを libfuzzuf.so なりでビルドして別のコードからAFLクラスを参照してもFrida modeが使えない）
いまはそれでもよいとして、必要となったらExecutorなりにfrida modeを入れるように改修するという選択肢もあります。
本PRではこの辺のスコープ設定どうしましょうか。

ちなみに、あおきはこういう素朴な実装で大目標を達成しちゃう話は好きです。

File: docs/frida_mode.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+## Build instruction
+
+To use Frida mode, specify `-DENABLE_FRIDA_TRACE=1` while generating a CMake's build system.  
+It will clone `AFLplusplus` submodule and build `afl-frida-trace.so` located under `AFLplusplus/frida_mode/` directory.
+
+```console
+$ cd /path/to/fuzzuf
+$ # Build plain (dynamically linked, uninstrumented) `exifutil` executable
+$ make -C docs/resources/exifutil clean all
+$ cmake -B build ... -DENABLE_FRIDA_TRACE=1
```
Comment (K-atc) [inline docs/frida_mode.md]:
```suggestion
$ cmake -B build . -DENABLE_FRIDA_TRACE=1
```
If the current directory hasn't changed after `cd /path/to/fuzzuf`, then I guess one dot is the right answer.

File: docs/frida_mode.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+## Build instruction
+
+To use Frida mode, specify `-DENABLE_FRIDA_TRACE=1` while generating a CMake's build system.  
+It will clone `AFLplusplus` submodule and build `afl-frida-trace.so` located under `AFLplusplus/frida_mode/` directory.
+
+```console
+$ cd /path/to/fuzzuf
+$ # Build plain (dynamically linked, uninstrumented) `exifutil` executable
+$ make -C docs/resources/exifutil clean all
```
Comment (K-atc) [inline docs/frida_mode.md:24]: It seems clang is required. But frida mode's documents does not mention that.
```
$ AFL_PATH=~/bin/AFL make -C docs/resources/exifutil clean all
make: ディレクトリ '/home/aoki/atla/fuzzuf-for-review/docs/resources/exifutil'　に入ります
rm -f *exifutil
cc -no-pie -fno-stack-protector -Wno-address-of-packed-member -O3 exifutil.c -o exifutil
cc -no-pie -fno-stack-protector -Wno-address-of-packed-member -O3 -static exifutil.c -o static-exifutil
cc -no-pie -fno-stack-protector -Wno-address-of-packed-member -O3 exifutil.c -o san-exifutil -fsanitize=address
afl-clang -no-pie -fno-stack-protector -Wno-address-of-packed-member -O3 exifutil.c -o afl-exifutil
afl-cc 2.57b by <lcamtuf@google.com>

[-] PROGRAM ABORT : Oops, failed to execute 'clang' - check your PATH
         Location : main(), afl-gcc.c:342

make: *** [Makefile:17: afl-exifutil] エラー 1
```

File: docs/resources/exifutil/Makefile [hunk 4]
```diff
@@ -1,12 +1,15 @@
 CC?=clang
 AFLCC?=afl-clang
-CFLAGS=-nopie -fno-stack-protector -Wno-address-of-packed-member -O3 -static
+CFLAGS=-no-pie -fno-stack-protector -Wno-address-of-packed-member -O3
```
Comment (K-atc) [inline docs/resources/exifutil/Makefile:3]: Without this change, we cannot compile exitfutil. Thanks!

File: docs/frida_mode.md (hunk 3 above)
Comment (K-atc) [reply docs/frida_mode.md:24]: Build fails when using clang-14:

```
$ AFL_PATH=~/bin/AFL CC=clang-14 make -C docs/resources/exifutil clean all
.... snipped ...
/tmp/.afl-3540237-1646036544.s:8760: Error: junk at end of line, first unrecognized character is `3'
/tmp/.afl-3540237-1646036544.s:8763: Error: file number less than one
/tmp/.afl-3540237-1646036544.s:8763: Error: junk at end of line, first unrecognized character is `3'
/tmp/.afl-3540237-1646036544.s:8772: Error: file number less than one
/tmp/.afl-3540237-1646036544.s:8772: Error: junk at end of line, first unrecognized character is `0'
/tmp/.afl-3540237-1646036544.s:8801: Error: file number less than one
/tmp/.afl-3540237-1646036544.s:8801: Error: junk at end of line, first unrecognized character is `1'
/tmp/.afl-3540237-1646036544.s:8822: Error: file number less than one
/tmp/.afl-3540237-1646036544.s:8822: Error: junk at end of line, first unrecognized character is `7'
/tmp/.afl-3540237-1646036544.s:8847: Error: file number less than one
/tmp/.afl-3540237-1646036544.s:8847: Error: junk at end of line, first unrecognized character is `1'
/tmp/.afl-3540237-1646036544.s:8875: Error: file number less than one
/tmp/.afl-3540237-1646036544.s:8875: Error: junk at end of line, first unrecognized character is `5'
/tmp/.afl-3540237-1646036544.s:8899: Error: junk at end of line, first unrecognized character is `"'
clang: error: assembler command failed with exit code 1 (use -v to see invocation)
make: *** [Makefile:17: afl-exifutil] エラー 1
```

**Review by K-atc — CHANGES_REQUESTED** [review]
I have no opinion except for the Build instruction.
* [x] (Supporting opinion) I understood that the fuzzuf's official Docker image does not include frida mode in the build. This is an experimental release, so I think it's fine to keep the current decision.
* [ ] I could **not** check if "Build instruction" is correct, because `make -C docs/resources/exifutil clean all` fails on my PC.
* [ ] Move submodule AFLplusplus under `third_party/AFLplusplus`

File: docs/frida_mode.md (hunk 2 above)
Comment (mkYYY) [reply docs/frida_mode.md]: I've put these three dots as a placeholder for other `cmake` options and that was a bad decision. Sorry for making you confused. (will fix in the next commit)

**Pull request merged**